File: backend/internal/media/media.go
// Package media is the vendor-neutral layer between music services and
// speakers. It exists so that "play this on those speakers" is one code path
// rather than one per bridge, and so the cross-vendor case — a KEF and a Sonos
// playing the same thing at once — has somewhere to live at all.
//
// The model is four nouns, documented in full in docs/MEDIA-PROTOCOL.md:
//
//	Provider  a music service (Spotify), which can search and can serve
//	          content over one or more routes
//	Endpoint  one speaker, addressed uniformly; adapters wrap the existing
//	          internal/sonos and internal/kef bridges
//	Zone      a named set of endpoints that play together, of any mix of
//	          vendors; a zone of one is the ordinary single-speaker case
//	Route     how content gets from a provider onto a zone
//
// Nothing here talks to hardware. The adapters do that, and they translate
// only — an adapter never emulates a capability the speaker lacks, because a
// capability that lies is worse than one that is absent: the route engine
// picks paths based on these declarations, and a wrong pick starts music in
// the wrong room.
package media

import (
	"context"
	"errors"
	"fmt"
	"time"
)

// Vendor identifies which bridge is behind an endpoint. It is deliberately
// coarse: the route engine uses it only to decide whether two endpoints can
// group natively with each other, never to special-case behaviour. Anything
// finer belongs in Capability.
type Vendor string

const (
	VendorSonos Vendor = "sonos"
	VendorKEF   Vendor = "kef"
)

// Capability is one thing an endpoint can do. Endpoints declare a set of
// these, and the route engine reasons entirely in terms of them — that is what
// lets a new bridge join without the engine learning about it.
//
// A capability describes the *hardware and its bridge*, not the current
// moment: a KEF in standby still has CapConnect, it just needs waking first.
// Conditions that change minute to minute are reported by Availability, not
// by dropping a capability.
type Capability uint32

const (
	// CapTransport is play/pause/next/previous. Every endpoint has it.
	CapTransport Capability = 1 << iota
	// CapVolume is 0-100 volume and mute.
	CapVolume
	// CapSeek is seeking within the current track.
	CapSeek
	// CapQueue is inspecting and mutating a queue of upcoming tracks.
	CapQueue
	// CapGroup is native multi-speaker grouping with same-vendor endpoints,
	// where the vendor's own clock keeps them in sync.
	CapGroup
	// CapPlayURI is being handed an arbitrary stream URL. This is the one
	// that makes the stream route possible, and it is the only capability
	// both vendors share for *starting* content.
	CapPlayURI
	// CapNativeService is streaming a music service directly, from an
	// account link held by the speaker itself. Sonos has this; KEF does not,
	// which is the asymmetry the whole route engine exists to paper over.
	CapNativeService
	// CapConnect is being targeted by a service's own cloud (Spotify
	// Connect). The inverse asymmetry: KEF has it, Sonos does not.
	CapConnect
	// CapWake is being woken from standby onto the network. Endpoints
	// without it are always reachable; endpoints with it must be woken
	// before any route that needs them to exist on the network.
	CapWake
)

// capNames drives Has, String and the JSON encoding. Kept in declaration
// order so a printed set reads consistently.
var capNames = []struct {
	cap  Capability
	name string
}{
	{CapTransport, "transport"},
	{CapVolume, "volume"},
	{CapSeek, "seek"},
	{CapQueue, "queue"},
	{CapGroup, "group"},
	{CapPlayURI, "play_uri"},
	{CapNativeService, "native_service"},
	{CapConnect, "connect"},
	{CapWake, "wake"},
}

// Has reports whether every capability in want is present. Passing a
// multi-bit want asks for all of them, which is what route predicates need.
func (c Capability) Has(want Capability) bool { return c&want == want }

// Names returns the set as stable strings, for the API and for error
// messages that have to tell a user what a speaker can actually do.
func (c Capability) Names() []string {
	var out []string
	for _, n := range capNames {
		if c&n.cap != 0 {
			out = append(out, n.name)
		}
	}
	return out
}

func (c Capability) String() string {
	names := c.Names()
	if len(names) == 0 {
		return "none"
	}
	return fmt.Sprint(names)
}

// MarshalJSON encodes a capability set as an array of names rather than an
// integer, so the frontend can check for a capability without a shared table
// of bit values that would drift.
func (c Capability) MarshalJSON() ([]byte, error) {
	names := c.Names()
	if names == nil {
		names = []string{}
	}
	// Hand-rolled: the names are a closed set of safe identifiers, so this
	// avoids pulling encoding/json into a hot path for no benefit.
	buf := make([]byte, 0, 16*len(names)+2)
	buf = append(buf, '[')
	for i, n := range names {
		if i > 0 {
			buf = append(buf, ',')
		}
		buf = append(buf, '"')
		buf = append(buf, n...)
		buf = append(buf, '"')
	}
	return append(buf, ']'), nil
}

// Descriptor is an endpoint's identity and what it can do — everything the
// route engine and the UI need without touching the device.
type Descriptor struct {
	ID     string     `json:"id"`
	Name   string     `json:"name"`
	Room   string     `json:"room,omitempty"`
	Vendor Vendor     `json:"vendor"`
	Model  string     `json:"model,omitempty"`
	Caps   Capability `json:"capabilities"`
	// GroupKey is the set within which CapGroup endpoints can natively
	// group. Same vendor is necessary but not sufficient — two Sonos
	// households on one LAN can't group with each other — so the adapter
	// decides what identifies a group domain rather than the engine
	// assuming vendor equality is enough.
	GroupKey string `json:"-"`
}

// PlayState is the normalised transport state. The bridges disagree on
// vocabulary (Sonos says PLAYING/PAUSED_PLAYBACK/STOPPED/TRANSITIONING, KEF
// says playing/paused/stopped) and both are preserved on their own endpoints;
// this is the common denominator the zone layer reports.
type PlayState string

const (
	StatePlaying       PlayState = "playing"
	StatePaused        PlayState = "paused"
	StateStopped       PlayState = "stopped"
	StateTransitioning PlayState = "transitioning"
)

// Track is the currently playing item, as much of it as the endpoint knows.
type Track struct {
	Title  string `json:"title,omitempty"`
	Artist string `json:"artist,omitempty"`
	Album  string `json:"album,omitempty"`
	// ArtURI may be absolute or relative to the endpoint. Sonos serves
	// artwork off the speaker itself (/getaa?...), so a relative value has
	// to be proxied by the caller — the same rule the sonos bridge already
	// documents, carried through rather than resolved here, because this
	// package has no HTTP client and should not grow one.
	ArtURI string `json:"art_uri,omitempty"`
}

// NowPlaying is one endpoint's live state, normalised.
type NowPlaying struct {
	State  PlayState `json:"state"`
	Volume int       `json:"volume"` // 0-100
	Muted  bool      `json:"muted"`
	Track  *Track    `json:"track,omitempty"`
	// Position and Duration are zero when the endpoint doesn't report them;
	// Duration is also zero for live streams, where there is nothing to
	// seek. Durations rather than the bridges' strings/millis so callers
	// stop reimplementing the parse.
	Position time.Duration `json:"-"`
	Duration time.Duration `json:"-"`
	// PositionMS/DurationMS are the wire form. Milliseconds because that is
	// what both the KEF bridge and the frontend already use.
	PositionMS int64 `json:"position_ms,omitempty"`
	DurationMS int64 `json:"duration_ms,omitempty"`
	// At is when this reading was taken, so a client extrapolating position
	// knows how stale its starting point is.
	At time.Time `json:"at"`
}

// SyncWire fills the millisecond fields from the Duration ones. Adapters set
// the typed fields and call this, so the two representations cannot drift.
func (n *NowPlaying) SyncWire() {
	n.PositionMS = n.Position.Milliseconds()
	n.DurationMS = n.Duration.Milliseconds()
}

// Playing reports whether audio is actually coming out, which is not the same
// as "not stopped" — transitioning is neither.
func (n *NowPlaying) Playing() bool { return n != nil && n.State == StatePlaying }

// Endpoint is one speaker. Implementations wrap a bridge and translate; see
// the package doc on why they must not emulate.
//
// Every method takes a context and performs device I/O. None of them may be
// called while store.Mu is held — that is the rule from CLAUDE.md, and the
// zone layer's staged execution is what enforces it.
type Endpoint interface {
	Descriptor() Descriptor
	State(ctx context.Context) (*NowPlaying, error)

	Play(ctx context.Context) error
	Pause(ctx context.Context) error
	Next(ctx context.Context) error
	Previous(ctx context.Context) error
	SetVolume(ctx context.Context, level int) error
	SetMute(ctx context.Context, muted bool) error
}

// The optional interfaces below correspond to capabilities. A caller that
// holds a capability may type-assert to the matching interface and expect it
// to succeed; adapters must keep the two in step. Declaring CapSeek without
// implementing Seeker is a bug in the adapter, and assertCaps catches it in
// tests.

// Seeker is CapSeek.
type Seeker interface {
	Seek(ctx context.Context, pos time.Duration) error
}

// QueueItem is one entry in an endpoint's queue.
type QueueItem struct {
	Position int    `json:"position"` // 1-based
	Title    string `json:"title,omitempty"`
	Artist   string `json:"artist,omitempty"`
	Album    string `json:"album,omitempty"`
	ArtURI   string `json:"art_uri,omitempty"`
	URI      string `json:"uri,omitempty"`
}

// Queuer is CapQueue.
type Queuer interface {
	Queue(ctx context.Context) ([]QueueItem, error)
	ClearQueue(ctx context.Context) error
}

// Grouper is CapGroup. Join and Leave address the vendor's native grouping;
// the route engine only ever calls them between endpoints sharing a GroupKey.
type Grouper interface {
	Join(ctx context.Context, coordinator Endpoint) error
	Leave(ctx context.Context) error
	// Coordinator reports the endpoint currently leading this one's group,
	// or "" when it stands alone. Used to avoid re-grouping speakers that
	// are already arranged correctly, which would interrupt playback.
	Coordinator(ctx context.Context) (string, error)
}

// Metadata is what a URI player should display while playing a stream. All
// fields are advisory: an endpoint shows what it can.
type Metadata struct {
	Title  string
	Artist string
	Album  string
	ArtURI string
	// Live marks a stream with no end, which changes how Sonos presents it
	// (no scrubber, no track duration) and stops it trying to advance.
	Live bool
}

// URIPlayer is CapPlayURI — the capability the stream route is built on.
type URIPlayer interface {
	PlayURI(ctx context.Context, uri string, meta Metadata) error
}

// NativeServicePlayer is CapNativeService: hand the speaker a service URI and
// let it stream from its own account link. The account token comes from the
// provider, which is why the argument is opaque here.
type NativeServicePlayer interface {
	// PlayNative starts service content. uri and metadata are whatever the
	// provider's NativeItem produced for this vendor — the media layer
	// carries them without interpreting them.
	PlayNative(ctx context.Context, uri, metadata string) error
	// ServiceAccount fetches the speaker's link for a named service, or an
	// error when the household has no such link.
	ServiceAccount(ctx context.Context, service string) (Account, error)
}

// Account is a speaker's link to a music service, opaque to this package and
// meaningful only to the provider that issued the lookup.
type Account struct {
	// SID and Serial are the Sonos shape (service id + account serial).
	// Kept concrete rather than as `any` because there is exactly one
	// vendor with native service links, and a fake generic type here would
	// be pretending to an extensibility that isn't real yet.
	SID    int
	Serial string
	Type   int
}

// ConnectTarget is CapConnect: this endpoint can be pointed at by a service's
// cloud. DeviceHint is what the endpoint believes it is called there, which
// the provider matches against its device list.
type ConnectTarget interface {
	// ConnectHint returns a pinned device id (empty when none) and the
	// names to match on, most specific first.
	ConnectHint() (deviceID string, names []string)
}

// Waker is CapWake.
type Waker interface {
	// Wake brings the endpoint out of standby and onto its network input.
	// Must be idempotent: waking an awake endpoint is a no-op, and routes
	// call it unconditionally rather than reading state first.
	Wake(ctx context.Context) error
}

// Errors the layer defines. Callers map these to status codes; the API
// package turns the "user can fix this" ones into 409s.
var (
	// ErrNoRoute means no route can serve this zone with this provider.
	// Always wrapped with a per-endpoint explanation — see RouteError.
	ErrNoRoute = errors.New("media: no route can play this to these speakers")
	// ErrUnknownEndpoint / ErrUnknownProvider are lookup failures.
	ErrUnknownEndpoint = errors.New("media: unknown endpoint")
	ErrUnknownProvider = errors.New("media: unknown provider")
	// ErrEmptyZone is a zone with no members, which is valid to store
	// (a user emptying a zone in the UI) but not to play to.
	ErrEmptyZone = errors.New("media: zone has no speakers")
)

File: frontend/src/lib/types/media.ts
// ── Media protocol ───────────────────────────────────────────────────────
// The vendor-neutral layer: speakers and services addressed uniformly, plus
// zones — sets of speakers that play together regardless of make. The types
// in sonos.ts and kef.ts stay: the per-speaker detail views need vendor
// specifics that don't generalise. See docs/MEDIA-PROTOCOL.md.

/** Which bridge is behind an endpoint. `airplay` is a protocol rather than a
 *  make — a RoPieee, an Apple TV and any shairport-sync box are all driven
 *  identically, which is exactly what a vendor means to the route engine. */
export type MediaVendor = "sonos" | "kef" | "airplay";

/**
 * One thing a speaker can do. These are the strings the backend emits, and
 * the UI keys off them rather than off the vendor — that is what lets a new
 * bridge appear without every component learning about it.
 */
export type MediaCapability =
  | "transport"
  | "volume"
  | "seek"
  | "queue"
  | "group"
  | "play_uri"
  | "native_service"
  | "connect"
  | "wake"
  /** Can be *pushed* audio rather than handed something to fetch. The
   *  inverse of play_uri in who holds the audio: HomeHub does the sending
   *  and keeps the clock. */
  | "airplay";

/** A speaker's identity and what it can do, without touching the device. */
export interface MediaEndpoint {
  id: string;
  name: string;
  room?: string;
  vendor: MediaVendor;
  model?: string;
  capabilities: MediaCapability[];
  /** Bridge-qualified id ("sonos:abc") — what a zone stores as a member. */
  member: string;
}

/** Normalised transport state, common across both bridges. */
export type MediaPlayState = "playing" | "paused" | "stopped" | "transitioning";

export interface MediaTrack {
  title?: string;
  artist?: string;
  album?: string;
  art_uri?: string;
  /** A live source describes itself differently — see SonosTrack. Only the
   *  Sonos bridge fills these; a KEF reports a title for everything. */
  stream?: string;
  station?: string;
}

export interface MediaNowPlaying {
  state: MediaPlayState;
  volume: number;
  muted: boolean;
  track?: MediaTrack;
  position_ms?: number;
  duration_ms?: number;
  /** When the reading was taken — extrapolate position from this, not now. */
  at: string;
}

/**
 * How content reaches a zone. Ordered best-first by the backend, and the
 * ordering is a guarantee: `stream` is only ever chosen when nothing else
 * can serve the whole zone, so a Sonos-only zone never lands on it.
 */
export type MediaRoute = "native" | "connect" | "group" | "airplay" | "stream";

/**
 * How well a route keeps speakers together.
 *
 * Three honest labels rather than a good/bad flag. `buffered` is the stream
 * route: a few hundred milliseconds apart, stable once running, and not the
 * sample-locked sync a native group gets. `clocked` is AirPlay: one sender,
 * one clock, every receiver told which sample belongs at which moment —
 * materially tighter than buffered, still not a vendor's own multi-room bus.
 */
export type MediaSync = "exact" | "single" | "buffered" | "clocked";

/** Whether something is usable right now, and if not what to do about it. */
export interface MediaAvailability {
  ok: boolean;
  configured: boolean;
  reason?: string;
}

export interface MediaProvider {
  id: string;
  name: string;
  availability: MediaAvailability;
  routes: MediaRoute[];
  /**
   * Whether the cross-vendor route works. A separate question from
   * `availability` — a service can search perfectly and be unable to stream,
   * and the two need different prompts.
   */
  streaming: MediaAvailability;
}

/** One member of a zone, with whatever live state it reported. */
export interface MediaZoneSpeaker extends MediaEndpoint {
  state?: MediaNowPlaying;
  /** Set when the speaker behind this member no longer exists. */
  missing?: boolean;
}

export interface MediaZone {
  id: string;
  name: string;
  /** Bridge-qualified speaker ids, in the order the user arranged them. */
  members: string[];
  room?: string;
  speakers: MediaZoneSpeaker[];
  /** What a play would use right now; absent when nothing can serve it. */
  route?: MediaRoute;
  sync?: MediaSync;
  /** Why that route, in words fit to show. Always set when route is. */
  reason?: string;
  /** Why nothing can serve this zone. Set instead of route/reason. */
  problem?: string;
  /** What a play here would actually sound like, source to speaker. */
  quality?: MediaChain;
}

// ── Quality ──────────────────────────────────────────────────────────────
// "Am I hearing lossless?" has no single answer: the audio passes through two
// hands — the service that encoded it and the path it took to the speaker —
// and a lossless path carrying a lossy source is not lossless. So the backend
// reports a chain with the weakest link named, and the UI shows that rather
// than a badge, because a badge would have to pick a lie.

export type MediaCodec = "pcm" | "alac" | "flac" | "vorbis" | "aac" | "mp3" | "";

export interface MediaQuality {
  codec?: MediaCodec;
  sample_rate?: number;
  bit_depth?: number;
  channels?: number;
  /** Absent for lossless and for unknown, which `lossless` tells apart. */
  bitrate_kbps?: number;
  lossless: boolean;
  /** Something HomeHub could not measure — a speaker's own session, whose
   *  parameters it never sees. Show as "up to", never as a measurement. */
  approximate?: boolean;
}

/** One link in the chain, with whatever is responsible for it named. */
export interface MediaQualityStage {
  name: string;
  quality: MediaQuality;
  detail?: string;
}

/** An improvement that is actually available. Absent is a real answer and
 *  must render as one rather than as a disabled button. */
export interface MediaQualityFix {
  /** The lever. "stream_quality" is the only value that maps to a control;
   *  empty means the improvement is something the listener does to their
   *  zone, so render it as a sentence and never as a button. */
  setting: string;
  label: string;
  detail?: string;
}

/** The end-to-end answer, in the states it actually has.
 *
 *  `up_to` is the one a boolean couldn't hold: on a route where the speaker
 *  holds the service account, nothing in the chain is lossy but nobody
 *  measured the far end either. Rendering it as "lossless" invents a reading;
 *  rendering it as "not lossless" tells someone with a lossless plan their
 *  system is worse than it is. It gets its own badge. */
export type MediaVerdict = "lossless" | "up_to" | "capped" | "unknown";

export interface MediaChain {
  source: MediaQualityStage;
  transport: MediaQualityStage;
  verdict: MediaVerdict;
  /** Bit-exact *and* known to be — true only for verdict `lossless`.
   *  Deliberately false for `up_to`; switch on `verdict` for the badge. */
  lossless: boolean;
  /** The stage that caps the result. What the UI leads with on `capped` —
   *  "not lossless" alone is not actionable. Empty on `up_to`: nothing is
   *  capping it, HomeHub just can't see the far end. */
  limited_by?: string;
  summary: string;
  fix?: MediaQualityFix;
}

/** How hard HomeHub's own decoder asks the service to compress. Household-wide
 *  because the decoder is one process holding one service session. */
export type StreamQuality = "best" | "balanced" | "saver";

export interface StreamQualityOption {
  value: StreamQuality;
  label: string;
  bitrate_kbps: number;
  /** What this choice costs and gains, in the terms the choice is made in. */
  detail: string;
}

/** One route and what it would sound like. */
export interface MediaRouteQuality {
  route: MediaRoute;
  label: string;
  /** True on the routes HomeHub decodes for — the only ones the setting
   *  reaches. Elsewhere the speaker holds the account and picks for itself. */
  decoded: boolean;
  chain: MediaChain;
}

export interface MediaProviderQuality {
  id: string;
  name: string;
  routes: MediaRouteQuality[];
}

/** GET /api/media/quality — the setting, the choices, and what every path
 *  through the house currently sounds like. */
export interface MediaQualityReport {
  stream_quality: StreamQuality;
  bitrate_kbps: number;
  options: StreamQualityOption[];
  providers: MediaProviderQuality[];
}

export type MediaItemKind = "track" | "album" | "playlist" | "artist" | "station";

export interface MediaItem {
  provider: string;
  kind: MediaItemKind;
  uri: string;
  title: string;
  subtitle?: string;
  art_uri?: string;
}

export interface MediaResults {
  tracks: MediaItem[];
  albums: MediaItem[];
  playlists: MediaItem[];
  artists: MediaItem[];
}

/** What a play answered: which route it took, and why. */
export interface MediaPlayResult {
  route: MediaRoute;
  sync: MediaSync;
  reason: string;
  /** Present only on the stream route. */
  stream_url?: string;
  speakers: string[];
  /** What it will sound like, source to speaker. */
  quality?: MediaChain;
}

/** One route that couldn't serve a zone, and the reason naming the speaker. */
export interface MediaRouteBlock {
  route: MediaRoute;
  reason: string;
}

export interface MediaZoneRoutes {
  speakers: string[];
  route?: MediaRoute;
  sync?: MediaSync;
  reason?: string;
  problem?: string;
  /** Every rejected route, in preference order. Present with `problem`. */
  blocked?: MediaRouteBlock[];
}

/**
 * One thing a room was asked to play, as HomeHub remembers it.
 *
 * This is the app's own memory rather than Spotify's, because Spotify's is
 * one list for the whole household and cannot say that the kitchen gets
 * radio at breakfast. `provider` decides how a tile replays it: "spotify"
 * goes back through the play path it came from, "sonos" is a household
 * favorite and is matched against the current favorites list by URI (a
 * favorite that has since been deleted simply stops appearing).
 */
export interface MediaPlay {
  provider: string;
  kind?: MediaItemKind;
  uri: string;
  title: string;
  sub?: string;
  art_uri?: string;
  /** What the room was called when this played. */
  room_name?: string;
  at: string;
  /**
   * How many times this room has started it. Absent on entries written
   * before the tally existed, which read as the one play that created them —
   * `playCount()` in lib/music/format.ts is the one place that decides so.
   */
  count?: number;
  /** When this room first played it — "since March" rather than "an hour
   *  ago". Absent on entries that predate the field. */
  first_at?: string;
  /** Plays by local hour, 24 slots. Absent on older entries, which is
   *  honest: they are not evidence about any hour in particular. */
  hours?: number[];
}

/** One room's play history, or the household's when it has none of its own. */
export interface MediaHistory {
  plays: MediaPlay[];
  /** True when these are the household's plays rather than this room's. */
  household: boolean;
}

/**
 * One track a room was heard playing — the other half of MediaPlay.
 *
 * A play is what someone *chose*: an album, a playlist, a station. This is
 * what came out of the speaker, written from what the speakers report rather
 * than from what HomeHub was asked to do. It is the only one of the two that
 * survives a queue being replaced, and the only one that knows about track
 * nine of something nobody picked by name.
 */
export interface HeardTrack {
  title: string;
  artist?: string;
  album?: string;
  art_uri?: string;
  /** The service URI where the source had one. Its absence is what makes a
   *  row a name rather than something to play again — radio, line-in and a
   *  KEF's own reporting all land here. */
  uri?: string;
  provider?: string;
  /** What the room was called when this played. */
  room_name?: string;
  at: string;
}

export interface MediaHeard {
  tracks: HeardTrack[];
  /** True when these are the household's tracks rather than this room's. */
  household: boolean;
}

/**
 * What a room keeps coming back to, rather than what it happened to play
 * last — and, when it has a habit at the hour it currently is, what it plays
 * *then*. The difference between offering the kitchen its breakfast radio at
 * eight in the morning and offering it last night's dinner record.
 */
export interface MediaTopPlays {
  plays: MediaPlay[];
  /** True when these are this room's habit at `hour`, false when they are
   *  its favourites overall. The shelf's label depends on it: a wall that
   *  says "you play this now" about a record this room has never played in
   *  the morning is exactly the confident wrongness §15 rules out. */
  by_hour: boolean;
  /** The local hour the ranking was asked for, 0–23. */
  hour: number;
}

/** One name and how many plays sit behind it. */
export interface ListeningTally {
  /** A room key for rooms, the artist's own line for artists — so a row can
   *  be acted on rather than only printed. */
  key: string;
  name: string;
  plays: number;
  /** The most recent play in this tally. */
  at: string;
}

/**
 * What the household's listening adds up to, summed over every room.
 *
 * Bounded by what the per-room lists still hold, so every number here is
 * "plays we remember" and never a lifetime total — `since` is what lets a
 * surface say which window it is describing instead of implying it covers
 * everything.
 */
export interface Listening {
  plays: number;
  /** How many distinct things those plays were. */
  items: number;
  /** Which rooms did the listening, busiest first. */
  rooms: ListeningTally[];
  /** The artist lines of the tracks and albums played, busiest first.
   *  Playlists and stations are left out — their second line is an owner or
   *  a service, and counting those would put "Spotify" at the top. */
  artists: ListeningTally[];
  /** The most-played items themselves, merged across rooms. */
  top: MediaPlay[];
  /** When the house listens: 24 slots, local time. */
  hours: number[];
  /** The oldest play still remembered. Absent when nothing has played. */
  since?: string;
}

// ── Music timers ─────────────────────────────────────────────────────────
// Music that starts and stops without anyone tapping anything: the half the
// socket scheduler could never reach. One type covers both uses because they
// differ only in which end of the fade they are on — arrive at 20 over ten
// minutes at 06:45, or take the room down to nothing in forty.

export type MusicTimerAction = "start" | "stop";

/** What a starting timer puts on. Carried in full rather than as a URI so
 *  06:45 is not the moment a catalog round trip has to succeed. */
export interface MusicTimerItem {
  provider?: string;
  kind?: string;
  uri?: string;
  title?: string;
  sub?: string;
  art_uri?: string;
}

export interface MusicTimer {
  id: string;
  name?: string;
  /** The media layer's destination key — "sonos:…", "kef:…", "zone:…" — the
   *  same vocabulary the play history uses. */
  room: string;
  action: MusicTimerAction;
  enabled: boolean;
  /** Set makes this a one-shot: it runs once and is deleted. This is what
   *  "sleep in forty minutes" is. */
  fires_at?: string;
  /** "HH:MM" plus days makes it recurring. Empty days means every day. */
  time?: string;
  days?: number[];
  item?: MusicTimerItem;
  /** Where the room ends up: the level to arrive at for a start, the level
   *  to fade down to for a stop. Absent leaves the volume alone. */
  volume?: number;
  /** How long to take getting there. Zero is a jump. */
  fade_minutes?: number;
  last_fired_at?: string;
}

/** A timer plus what the backend already knows and the wall would otherwise
 *  have to work out for itself. */
export interface MusicTimerView extends MusicTimer {
  /** What the house calls the room now — not necessarily what it was called
   *  when the timer was set. */
  room_name: string;
  /** When this next fires, so a row can say "in 6 hours" without
   *  reimplementing weekday arithmetic. Absent when it never will. */
  next_at?: string;
  /** A ramp is walking this room right now — the state between "sleep timer
   *  set" and "room quiet", otherwise invisible except as volume drifting
   *  on its own. */
  fading: boolean;
}

/** What setting a sleep timer answered. */
export interface MusicSleepResult {
  timer: MusicTimer;
  /** When the room actually goes quiet — the number worth reading back,
   *  rather than when the fade starts. */
  quiet_at: string;
}

/** One room an announcement could be sent to. */
export interface AnnounceRoom {
  id: string;
  name: string;
}

/** What the announcement control needs before anyone taps it. */
export interface AnnounceStatus {
  /** False when no speaker is answering — there is nowhere to announce to. */
  available: boolean;
  /** The rooms that would hear it, in the order they will be addressed. */
  rooms: AnnounceRoom[];
  /** False when no voice service is configured: a chime, and no words. */
  voice: boolean;
  max_text: number;
  /** The sentences the wall offers before its box, from household settings
   *  (edited in Settings, only read by the kiosk). They come from the server
   *  rather than a constant here for two reasons pulling the same way:
   *  typing is the worst thing a wall asks anyone to do, so the presets are
   *  most of what the control is — and they are read out by a voice that
   *  speaks one language, which the household picks. An empty list is a
   *  household that wants the box and nothing above it. */
  presets: string[];
}

/** What an announcement did. */
export interface AnnounceResult {
  rooms: string[];
  /** Rooms that didn't take it, so the surface can be honest about reach. */
  unreachable: string[];
  /** False when the words were dropped and only the chime played. */
  spoken: boolean;
  duration_ms: number;
}
